docs/features/service_history.md

```diff
@@ -0,0 +1,190 @@
+
+## Service History Generation
+
+The `ServiceHistoryEnrollment` and its associated `ServiceHistoryService` records are generated by a background task. This process is idempotent and can be re-run safely. The core logic resides in the `rebuild_service_history!` method within the `GrdaWarehouse::Tasks::ServiceHistory::Enrollment` class.
+
+The generation logic differs based on the program type:
+
+-   **Night-by-Night (NBN) Programs:** A `ServiceHistoryService` record is created for each specific service date recorded in the source data (e.g., a bed night service).
+-   **Entry/Exit Programs:** For programs like Entry/Exit shelters, Transitional Housing, and Permanent Housing, the system builds `ServiceHistoryService` records for each day from the client's entry date through the day before their exit date. When the entry and exit dates are the same we include that day. For ongoing enrollments, we extend the services through the latest date covered by the source export (or today, if the enrollment is managed in the operational HMIS).
+-   **Street Outreach & Contact Extrapolation:** Street Outreach projects can create synthetic daily services from contact records. When the `so_day_as_month` configuration flag is enabled—or when a project is configured to extrapolate contacts—we expand recorded contacts to fill the remainder of the month so that downstream reporting sees continuous engagement.
+
+## Storage & Performance Considerations
+
+ The `service_history_services` dataset is among the largest tables in the warehouse regardless of installation scale. To keep inserts fast and maintenance manageable, we physically shard it by year. PostgreSQL trigger logic (see `db/warehouse_structure.sql`) routes each new service day into the correct annual partition (`service_history_services_2000`, `service_history_services_2001`, …, `service_history_services_2050`) with an overflow table (`service_history_services_remainder`) for out-of-range data. Queries continue to reference the parent `service_history_services` relation, but PostgreSQL prunes to the relevant partitions based on date filters.
+
+For analytic workloads that benefit from a consistent snapshot—and to avoid the cross-partition overhead of per-client lookups—we expose the `service_history_services_materialized` relation. The `GrdaWarehouse::ServiceHistoryServiceMaterialized` model manages that materialized view, including `refresh!` (to repopulate it) and `rebuild!` (to drop, recreate, and reindex). A notifier-backed `double_check_materialized_view` helper compares recent homeless dates between the live partitions and the materialized copy so we can detect drift. Future work (see ADR 0006) explores replacing daily rows with period-based ranges, which could eliminate the need for this materialized view entirely.
```

Comment:
Might be worth mentioning that there are some guard rails in place so we don't generate service back to the year 100 (we used to!) or far into the future.